### PR TITLE
【PIR API adaptor No.124】 optimizer.Lamb

### PR DESCRIPTION
**Pull request by Liyulingyue**
### PR types

Others
### PR changes

APIs 
### Description

https://github.com/PaddlePaddle/Paddle/issues/58067 124

### CODE REVIEW

File: test/legacy_test/test_imperative_optimizer_v2.py [hunk 1]
```diff
@@ -783,6 +784,7 @@ def get_optimizer_dygraph(self, parameter_list):
         )
         return optimizer
 
+    @test_with_pir_api
```
Comment (MarioLulab) [inline test/legacy_test/test_imperative_optimizer_v2.py]: 应该修饰 TestImperativeLambOptimizer._test_lamb ，并且将 _test_lamb 修改为 test_lamb

File: test/legacy_test/test_lambv2_op.py [hunk 2]
```diff
@@ -200,6 +202,7 @@ def test_lamb_op(self):
 
 
 class TestLambOpMultiPrecision(unittest.TestCase):
+    @test_with_pir_api
```
Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py]: 应该去装饰 test_main，因为 test_main 里才对结果做比较

**Comment by MarioLulab** [comment]
还貌似遗漏了 test/legacy_test/test_lamb_op.py 的单测适配？

**Review by YuanRisheng — COMMENTED** [review]
image
这里新ir并没有适配这个未来会废弃掉的接口，可以尝试修改使用paddle.nn.Linear接口

File: test/legacy_test/test_lambv2_op.py [hunk 3]
```diff
@@ -124,7 +126,7 @@ def _build_static_model(main, startup, seed=100):
                     name='X', shape=[-1, 13], dtype='float32'
                 )
                 y = paddle.static.data(name='Y', shape=[-1, 1], dtype='float32')
-                prediction = paddle.static.nn.fc(x, size=1, activation=None)
+                prediction = paddle.nn.Linear.fc(x, size=1, activation=None)
```
Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py]:
```suggestion
                linear = paddle.nn.Linear(in_features=x.shape[-1], out_features=1)
                prediction = linear(x)
```

File: test/legacy_test/test_imperative_optimizer_v2.py
Comment (MarioLulab) [inline test/legacy_test/test_imperative_optimizer_v2.py:1]: ci 还挂在了这里：
image

可以将这里的 base.Program() 改为 static.Program()
base.program_guard 改成 static.program_guard

File: test/legacy_test/test_imperative_base.py [hunk 4]
```diff
@@ -14,15 +14,15 @@
...
-    scope = scope if scope else base.core.Scope()
-    with base.scope_guard(scope):
-        with base.program_guard(prog, startup_prog):
-            with base.unique_name.guard():
+    prog = main if main else static.Program()
+    startup_prog = startup if startup else static.Program()
+    scope = scope if scope else static.core.Scope()
+    with static.scope_guard(scope):
+        with static.program_guard(prog, startup_prog):
+            with static.unique_name.guard():
```
Comment (MarioLulab) [inline test/legacy_test/test_imperative_base.py]: 不需要所有 base 都改，除 Program 和 program_guard 外的都应该需要复原，否则会导致其他的单测没法通过

```suggestion
    prog = main if main else static.Program()
    startup_prog = startup if startup else static.Program()
    scope = scope if scope else base.core.Scope()
    with static.scope_guard(scope):
        with static.program_guard(prog, startup_prog):
            with base.unique_name.guard():
```

**Comment by MarioLulab** [comment]
需要 pre-commit 一下

File: test/legacy_test/test_lambv2_op.py [hunk 5]
```diff
@@ -124,7 +126,8 @@ def _build_static_model(main, startup, seed=100):
                     name='X', shape=[-1, 13], dtype='float32'
                 )
                 y = paddle.static.data(name='Y', shape=[-1, 1], dtype='float32')
-                prediction = paddle.static.nn.fc(x, size=1, activation=None)
+                linear = paddle.nn.Linear(in_features=x.shape[-1], out_features=1)
```
Comment (Liyulingyue) [inline test/legacy_test/test_lambv2_op.py]:
```suggestion
                linear = paddle.nn.Linear(
                    in_features=x.shape[-1], out_features=1
                )
```

File: test/legacy_test/test_lambv2_op.py
Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py:1]: 这几行需要适配 pir
image

1. `x = main_prog.global_block().var(x.name)` 删除
2. `if x.dtype == core.VarDesc.VarType.FP16` 改为 if x.dtype in (core.VarDesc.VarType.FP16, core.DataType.FLOAT16)

Comment (MarioLulab) [reply test/legacy_test/test_lambv2_op.py:1]: 在 pir 模式下，当前场景 OpResult 无法从 python 侧直接获取 name。所以这里需要做修改
image

`feed_dict = {x.name: x_np}` 改为 `feed_dict = {'x' : x_np}`

Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py:1]: image

fetch_list = [avg_loss]

image

fetch_list = [loss]

Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py:1]: test_lamb_op_with_multi_steps 单测下的所有 base.Program() 麻烦改成 paddle.static.Program()

File: test/legacy_test/test_imperative_base.py [hunk 6]
```diff
@@ -14,15 +14,15 @@
 
 import contextlib
 
-from paddle import base
+from paddle import static
```
Comment (MarioLulab) [inline test/legacy_test/test_imperative_base.py]:
```suggestion
from paddle import static, base
```

File: test/legacy_test/test_imperative_optimizer_v2.py [hunk 7]
```diff
@@ -790,7 +791,8 @@ def get_optimizer(self):
         return optimizer
 
     # should fix: may fail in CI-windows
-    def _test_lamb(self):
+    @test_with_pir_api
+    def test_lamb(self):
```
Comment (MarioLulab) [inline test/legacy_test/test_imperative_optimizer_v2.py]:
```suggestion
    def _test_lamb(self):
```

这行 CI-windows 会出问题，还是复原吧，此 pr 不做修复

**Review by MarioLulab — COMMENTED** [review]
nice work
test_lambv2_op 的单测需要使用到 linalg.py 里的 vector_norm，辛苦顺便适配一下 pir 吧

image

File: test/legacy_test/test_lambv2_op.py [hunk 8]
```diff
@@ -256,7 +260,7 @@ def get_parameter(var):
 
             weight_np, bias_np = None, None
             for i in range(n):
-                feed_dict = {x.name: x_np}
+                feed_dict = {'x' : x_np}
```
Comment (Liyulingyue) [inline test/legacy_test/test_lambv2_op.py]:
```suggestion
                feed_dict = {'x':x_np}
```

**Comment by MarioLulab** [comment]
https://github.com/PaddlePaddle/Paddle/blob/85c57dabb6a372de94a41db84020a97f20a9df12/python/paddle/tensor/linalg.py#L431
这一行的 in_dynamic_mode() 替换为 in_dynamic_or_pir_mode() 看是否能通过 ci ？

File: test/legacy_test/test_lambv2_op.py [hunk 9]
```diff
@@ -266,6 +270,7 @@ def get_parameter(var):
                 np.testing.assert_array_equal(bias_np, get_parameter(bias))
             return weight_np, bias_np
 
+    @test_with_pir_api
```
Comment (MarioLulab) [inline test/legacy_test/test_lambv2_op.py]: 先取消此单测，因为混合精度单测涉及 class `OptimizerWithMixedPrecision`，该 class 暂不支持 pir 模式。请在 pr 描述里补充说明一下这个情况，并更新单测覆盖率~

**Review by MarioLulab — COMMENTED** [review]
nice work~ 但是遇到了不兼容的问题，需要补充到 pr 描述里

File: test/legacy_test/test_lambv2_op.py (hunk 9 above)
Comment (Liyulingyue) [inline test/legacy_test/test_lambv2_op.py]:
```suggestion
```

File: test/legacy_test/test_lambv2_op.py [hunk 10]
```diff
@@ -256,7 +260,7 @@ def get_parameter(var):
 
             weight_np, bias_np = None, None
             for i in range(n):
-                feed_dict = {x.name: x_np}
+                feed_dict = {'x':x_np}
```
Comment (Liyulingyue) [inline test/legacy_test/test_lambv2_op.py]:
```suggestion
                feed_dict = {'x': x_np}
```

**Comment by MarioLulab** [comment]
需要 pre-commit 一下

File: test/legacy_test/test_imperative_base.py [hunk 11]
```diff
@@ -14,15 +14,15 @@
 
 import contextlib
 
-from paddle import base
+from paddle import static, base
```
Comment (Liyulingyue) [inline test/legacy_test/test_imperative_base.py]:
```suggestion
from paddle import base, static
```

**Review by MarioLulab — APPROVED** [review]
LGTM
lamb 的混合精度单测涉及 class OptimizerWithMixedPrecision，该 class 暂不支持 pir 模式。请在 pr 描述里补充说明一下这个情况，并更新单测覆盖率~

**Review by 0x45f — APPROVED** [review]
辛苦更新下单测覆盖率~
